cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -3541,6 +3649,24 @@ TEST(TestQuantileKernel, AllNullsOrNaNs) {
   }
 }
 
+TEST(TestQuantileKernel, Decimal) {
+  auto check = [](const std::shared_ptr<Array>& input, QuantileOptions options,
+                  const std::shared_ptr<Array>& expected) {
+    ASSERT_OK_AND_ASSIGN(Datum out, Quantile(input, options));
+    auto out_array = out.make_array();
+    ValidateOutput(*out_array);
+    AssertArraysEqual(*expected, *out_array, /*verbose=*/true);
+  };
+  for (const auto& ty : {decimal128(3, 2), decimal256(3, 2)}) {
+    check(ArrayFromJSON(ty, R"(["1.00", "5.00", null])"),
+          QuantileOptions(0.5, QuantileOptions::LINEAR),
+          ArrayFromJSON(float64(), R"([3.00])"));
```

Comment:
Hmm, should LINEAR return the input decimal type instead of float64?

Reply:
We'd then have to convert the quantile `q` to decimal (and accept truncation or promote the precision in the output type). IMO, it's more consistent for interpolated quantiles to always be a float.

Reply:
Ah, fair enough. No choice will be perfect anyway.